### PR TITLE
Add handling for whatsapp interactive message throught quick reply

### PR DESCRIPTION
**Pull request by rasoro**
This allow Courier to handle quickreply message from RapidPro and send as interactive message. And also allow Rapidpro to handle interactive message responses by interpreting the button text or the option title selected in list as message text.

More details: #360

### CODE REVIEW

File: handlers/whatsapp/whatsapp.go [hunk 1]
```diff
@@ -593,25 +658,106 @@ func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStat
...
 
-				// add logs to our status
-				for _, log := range logs {
-					status.AddLog(log)
+			qrs := msg.QuickReplies()
+
+			channel := msg.Channel()
+
+			wppVersion := channel.ConfigForKey("version", "0").(string)
+			hasCompatibleVersion := wppVersion >= interactiveMsgMinSupVersion
```
Comment (nicpottier) [inline handlers/whatsapp/whatsapp.go:667]: This is not going to work for string versions. You'll need to use something like the semantic versioning package `semver`

Comment (nicpottier) [reply handlers/whatsapp/whatsapp.go:667]: Also we need the RapidPro side of this as well right?

File: handlers/whatsapp/whatsapp.go [hunk 2]
```diff
@@ -593,25 +658,106 @@ func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStat
...
+						status.AddLog(log)
+					}
+					if err != nil {
+						break
+					}
+
+					// if this is our first message, record the external id
+					if i == 0 {
+						status.SetExternalID(externalID)
+					}
```
Comment (nicpottier) [inline handlers/whatsapp/whatsapp.go:760]: A lot of repetition above, can you refactor this so we aren't repeating `sendWhatsAppMsg`, logging and external id over and over? Maybe it's a callback function that's passed in to build the payload?

**Review by nicpottier — CHANGES_REQUESTED** [review]

File: handlers/whatsapp/whatsapp.go (hunk 1 above)
Comment (rasoro) [reply handlers/whatsapp/whatsapp.go]: Hello, nicpottier. To improve version checking i used a semver package in the new commit as you recommended. Does it look better that way?

File: handlers/whatsapp/whatsapp.go (hunk 2 above)
Comment (rasoro) [reply handlers/whatsapp/whatsapp.go]: I think your suggestion makes a lot of sense. So I tried to improve it that way, what did you think of how it is now? any suggestions are welcome and i will always be willing to try to improve.

File: handlers/whatsapp/whatsapp.go [hunk 3]
```diff
@@ -33,6 +34,8 @@ const (
 	channelTypeTXW = "TXW"
 
 	mediaCacheKeyPattern = "whatsapp_media_%s"
+
+	interactiveMsgMinSupVersion = "v2.35.2"
```
Comment (nicpottier) [inline handlers/whatsapp/whatsapp.go:38]: might as well spell out `Supported` here.

**Review by nicpottier — APPROVED** [review]
Definitely looks better, thanks! Is there a rapidpro side of this yet to populate the version?

**Comment by norkans7** [comment]
I think on this we need to have the channel work right away once we add it.
Currently the channel will be broken until the version is synced with the templates.

I think the interactive messages can be the default behavior for WA channel type but have a migration on RapidPro to set a flag(my preference) or the version to the current channels and then new channels without that set are considered compatible

**Comment by matmsa27** [comment]
> I think on this we need to have the channel work right away once we add it.
> Currently the channel will be broken until the version is synced with the templates.

Yep, I agree with you @norkans7. To merge this is necessary the [task for template message](https://github.com/nyaruka/rapidpro/pull/3364) get the version from the WhatsApp instance and set the version on channel config for each channel.

> I think the interactive messages can be the default behavior for WA channel type but have a migration on RapidPro to set a flag(my preference) or the version to the current channels and then new channels without that set are considered compatible

For each channel created, the task for sync templates will do the job to set the version for each channel. To new channels, what do you think do an adjustment on creating a channel, to get the version on WhatsApp instance and set key `version` on channel config?

**Comment by norkans7** [comment]
I think as this feature will be supported going forward, it is best the channel that is added without the version set to work with those feature enabled.

And a migration on RapidPro to set the version on the existing channels to set the version so we skip the unsupported features. So the use of the flag of version is temporal as we migrate the channels to the latest version

the task to sync the version will take a while to populate the version as of https://github.com/nyaruka/rapidpro/pull/3364

**Comment by nicpottier** [comment]
@rasoro If you can fix the conflicts we're ready to merge this now that the RapidPro side is done.

**Comment by rasoro** [comment]
hi, @nicpottier . Thanks for letting me know, fixed.

**Comment by nicpottier** [comment]
Thanks!

**Pull request merged**